toolchain/check/cpp_type_mapping.cpp

```diff
@@ -0,0 +1,183 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "toolchain/check/cpp_type_mapping.h"
+
+#include <cstddef>
+#include <iostream>
+#include <optional>
+
+#include "clang/AST/Type.h"
+#include "clang/Basic/TargetInfo.h"
+#include "toolchain/base/int.h"
+#include "toolchain/base/kind_switch.h"
+#include "toolchain/base/value_ids.h"
+#include "toolchain/check/context.h"
+#include "toolchain/check/convert.h"
+#include "toolchain/check/literal.h"
+#include "toolchain/sem_ir/class.h"
+#include "toolchain/sem_ir/ids.h"
+#include "toolchain/sem_ir/inst.h"
+#include "toolchain/sem_ir/type.h"
+#include "toolchain/sem_ir/typed_insts.h"
+
+namespace Carbon::Check {
+
+// Find the bit width of an integer literal.
+// The default bit width is 32. If the literal's bit width is greater than 32,
+// the bit width is increased to 64.
+static auto FindIntLiteralBitWidth(Context& context, SemIR::InstId arg_id)
+    -> IntId {
+  auto arg_const_id = context.constant_values().Get(arg_id);
+  if (!arg_const_id.is_constant() ||
+      arg_const_id == SemIR::ErrorInst::ConstantId ||
+      arg_const_id.is_symbolic()) {
+    // TODO: Add tests for these cases.
+    return IntId::None;
+  }
+  auto arg = context.insts().GetAs<SemIR::IntValue>(
+      context.constant_values().GetInstId(arg_const_id));
+  llvm::APInt arg_val = context.ints().Get(arg.int_id);
+  unsigned arg_non_sign_bits = arg_val.getSignificantBits() - 1;
+
+  // TODO: What if the literal is larger than 64 bits? Currently an error is
+  // reported that the int value is too large for type `i64`. Maybe try to fit
+  // in i128/i256? Try unsigned?
+  return (arg_non_sign_bits <= 32) ? IntId::MakeRaw(32) : IntId::MakeRaw(64);
+}
+
+// Maps a Carbon builtin type to a Cpp type. Returns an empty QualType if the
+// type is not supported.
+// TODO: Have both Carbon -> C++ and C++ -> Carbon mappings in a single place
+// to keep them in sync.
+static auto TryMapBuiltinType(Context& context, SemIR::InstId inst_id,
+                              SemIR::TypeId type_id) -> clang::QualType {
+  // TODO: Object representation should not be used here. Instead, use
+  // NumericTypeLiteralInfo to decompose a ClassType into a uN / iN / fN type.
+  auto object_repr_id = context.sem_ir().types().GetObjectRepr(type_id);
+  if (!object_repr_id.has_value()) {
+    return clang::QualType();
+  }
+  auto type_inst_id = context.sem_ir().types().GetInstId(object_repr_id);
+  auto inst = context.insts().Get(type_inst_id);
```

Comment:
Should we merge these lines and call it `type_inst`?

```suggestion
  auto type_inst = context.insts().Get(context.sem_ir().types().GetInstId(object_repr_id));
```

Reply:
Done.